# ASSOCIATION RULE MINING USING PHYTON

In [1]:
!pip install apyori

  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5975 sha256=c77e883dbd5e538002c721a6a9a38061823139d0f0ad6a76f85357b0bbcc5823
  Stored in directory: c:\users\user\appdata\local\pip\cache\wheels\1b\02\6c\a45230be8603bd95c0a51cd2b289aefdd860c1a100eab73661
Successfully built apyori


In [3]:
import numpy as np
import pandas as pd
from apyori import apriori

In [5]:
data=pd.read_csv("store_data.csv",header=None)

In [6]:
data

0                  1            2                 3   \
0            shrimp            almonds      avocado    vegetables mix   
1           burgers          meatballs         eggs               NaN   
2           chutney                NaN          NaN               NaN   
3            turkey            avocado          NaN               NaN   
4     mineral water               milk   energy bar  whole wheat rice   
...             ...                ...          ...               ...   
7496         butter         light mayo  fresh bread               NaN   
7497        burgers  frozen vegetables         eggs      french fries   
7498        chicken                NaN          NaN               NaN   
7499       escalope          green tea          NaN               NaN   
7500           eggs    frozen smoothie  yogurt cake    low fat yogurt   

                4                 5     6               7             8   \
0     green grapes  whole weat flour  yams  cottage cheese  energy drink   
1              NaN               NaN   NaN             NaN           NaN   
2              NaN               NaN   NaN             NaN           NaN   
3              NaN               NaN   NaN             NaN           NaN   
4        green tea               NaN   NaN             NaN           NaN   
...            ...               ...   ...             ...           ...   
7496           NaN               NaN   NaN             NaN           NaN   
7497     magazines         green tea   NaN             NaN           NaN   
7498           NaN               NaN   NaN             NaN           NaN   
7499           NaN               NaN   NaN             NaN           NaN   
7500           NaN               NaN   NaN             NaN           NaN   

                9               10         11     12     13             14  \
0     tomato juice  low fat yogurt  green tea  honey  salad  mineral water   
1              NaN             NaN        NaN    NaN    NaN            NaN   
2              NaN             NaN        NaN    NaN    NaN            NaN   
3              NaN             NaN        NaN    NaN    NaN            NaN   
4              NaN             NaN        NaN    NaN    NaN            NaN   
...            ...             ...        ...    ...    ...            ...   
7496           NaN             NaN        NaN    NaN    NaN            NaN   
7497           NaN             NaN        NaN    NaN    NaN            NaN   
7498           NaN             NaN        NaN    NaN    NaN            NaN   
7499           NaN             NaN        NaN    NaN    NaN            NaN   
7500           NaN             NaN        NaN    NaN    NaN            NaN   

          15                 16               17       18         19  
0     salmon  antioxydant juice  frozen smoothie  spinach  olive oil  
1        NaN                NaN              NaN      NaN        NaN  
2        NaN                NaN              NaN      NaN        NaN  
3        NaN                NaN              NaN      NaN        NaN  
4        NaN                NaN              NaN      NaN        NaN  
...      ...                ...              ...      ...        ...  
7496     NaN                NaN              NaN      NaN        NaN  
7497     NaN                NaN              NaN      NaN        NaN  
7498     NaN                NaN              NaN      NaN        NaN  
7499     NaN                NaN              NaN      NaN        NaN  
7500     NaN                NaN              NaN      NaN        NaN  

[7501 rows x 20 columns]

#### NaN dalam data ini bukanlah Missing Data tetapi memang tidak ada transaksi yang terjadi

In [8]:
records=[]
for i in range(0,7501): records.append([str(data.values[i,j]) for j in range (0,20) if str(data.values[i,j]) !='nan'])

## Applying Apriori

In [9]:
min_purchased=35
support=min_purchased/7500
support

0.004666666666666667

#### Note: Terdapat 4 parameter dalam menggunakan apriori, data dari record, minimum support (35) bergantung pada ketentuan dari peneliti, minimum confidence (20%), dan lift

#### Lift adalah rasio. Lift (A->B) meningkatkan penjualan B jika A dibeli. Jika Lift=1 maka tidak ada hubungan antara A dan B. Lift >1 maka jika A dibeli maka B juga dibeli. Lift<1 kedua produk tidak dibeli bersamaan.

#### min_length: kita menentuka berapa rule yang akan digunakan (2) A dan B bergantung pada peneliti

In [12]:
association_rules=apriori(records, min_support=support, min_confidence=0.2, min_lift=3, min_length=2)
association_results=list(association_rules)

In [13]:
print(len(association_results))

22


In [14]:
print(association_results[0])

RelationRecord(items=frozenset({'escalope', 'mushroom cream sauce'}), support=0.005732568990801226, ordered_statistics=[OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)])


#### Nilai confidence 30.07% artinya 30.07% transaksi yang mengandung mushroom sauce juga mengandung escalope. Lift>1, maka orang yang membeli mushroom juga membeli escalope.

In [15]:
for item in association_results:
    pair=item[0]
    items=[x for x in pair]
    print("Rules: "+items[0]+"->"+items[1])
    print("Support: "+ str(item[1]))
    print("Cofindence: "+ str(item[2][0][2]))
    print("Lift: "+ str(item[2][0][3]))
    print("======================================")

Rules: escalope->mushroom cream sauce
Support: 0.005732568990801226
Cofindence: 0.3006993006993007
Lift: 3.790832696715049
Rules: pasta->escalope
Support: 0.005865884548726837
Cofindence: 0.3728813559322034
Lift: 4.700811850163794
Rules: ground beef->herb & pepper
Support: 0.015997866951073192
Cofindence: 0.3234501347708895
Lift: 3.2919938411349285
Rules: ground beef->tomato sauce
Support: 0.005332622317024397
Cofindence: 0.3773584905660377
Lift: 3.840659481324083
Rules: olive oil->whole wheat pasta
Support: 0.007998933475536596
Cofindence: 0.2714932126696833
Lift: 4.122410097642296
Rules: pasta->shrimp
Support: 0.005065991201173177
Cofindence: 0.3220338983050847
Lift: 4.506672147735896
Rules: chocolate->shrimp
Support: 0.005332622317024397
Cofindence: 0.23255813953488375
Lift: 3.2545123221103784
Rules: ground beef->cooking oil
Support: 0.004799360085321957
Cofindence: 0.5714285714285714
Lift: 3.2819951870487856
Rules: ground beef->spaghetti
Support: 0.008665511265164644
Cofindence: 0.

#### Rules: Olive Oil -> Soup. Transaksi yang mengandung olive oil dari keseluruhan transaksi memiliki rasio sebesar 0.00519.  Nilai confidence 22.5% artinya 22.5% transaksi yang mengandung olive oil juga mengandung soup. lift>1 artinya maka orang yang membeli olive oil juga membeli soup.

#### Rules: Olive Oil -> Spaghetti. Transaksi yang mengandung olive oil dari keseluruhan transaksi memiliki rasio sebesar 0.00506.  Nilai confidence 20.1% artinya 20.1% transaksi yang mengandung olive oil juga mengandung spaghetti. lift>1 artinya maka orang yang membeli olive oil juga membeli sapghetti. 